File: scripts/gemini-doctor.py
import os
import subprocess
import json

GREEN = '\033[92m'
RED = '\033[91m'
RESET = '\033[0m'

def print_ok(msg):
    print(f"{GREEN}{msg} [OK]{RESET}")

def print_bad(msg):
    print(f"{RED}{msg} [NOT IN GOOD SHAPE]{RESET}")

def check_node_version():
    print("Checking Node.js version...")
    required = (20, 0, 0)
    try:
        version = subprocess.check_output(["node", "-v"]).decode().strip().lstrip('v')
        major, minor, patch = map(int, version.split('.'))
        if (major, minor, patch) < required:
            print_bad(f"Node.js version {version} is too old. Required: >=20.0.0")
        else:
            print_ok(f"Node.js version {version}")
    except Exception as e:
        print_bad(f"Node.js not found: {e}")

def check_cli_version():
    print("\nChecking Gemini CLI version...")
    try:
        with open("package.json") as f:
            data = json.load(f)
            version = data.get("version")
            if version:
                print_ok(f"Gemini CLI version {version} is OK.")
            else:
                print_bad("Gemini CLI version not found in package.json.")
    except Exception as e:
        print_bad(f"Could not read package.json: {e}")

def check_gcloud_auth():
    print("\nChecking gcloud authentication...")
    adc_path = os.environ.get("GOOGLE_APPLICATION_CREDENTIALS")
    if adc_path and os.path.exists(adc_path):
        print_ok(f"ADC file found: {adc_path}")
    else:
        print_bad("ADC file not found or GOOGLE_APPLICATION_CREDENTIALS not set.")
    try:
        subprocess.check_output(["gcloud", "auth", "application-default", "print-access-token"], stderr=subprocess.STDOUT)
        print_ok("gcloud ADC authentication")
    except Exception as e:
        print_bad(f"gcloud ADC authentication failed: {e}")

def check_env():
    print("\nChecking environment variables...")
    for var in ["GOOGLE_CLOUD_PROJECT", "GOOGLE_CLOUD_LOCATION"]:
        if os.environ.get(var):
            print_ok(f"{var} is set: {os.environ[var]}")
        else:
            print_bad(f"{var} is NOT set.")

def check_api_endpoint():
    print("\nChecking API endpoint configuration...")
    gemini_api_key = os.environ.get("GEMINI_API_KEY")
    google_api_key = os.environ.get("GOOGLE_API_KEY")
    google_cloud_project = os.environ.get("GOOGLE_CLOUD_PROJECT")
    google_cloud_location = os.environ.get("GOOGLE_CLOUD_LOCATION")

    auth_type = None
    if gemini_api_key:
        auth_type = "USE_GEMINI"
        print("  + Using Gemini API key.")
        print("  - Endpoint: generativelanguage.googleapis.com")
    elif google_api_key or (google_cloud_project and google_cloud_location):
        auth_type = "USE_VERTEX_AI"
        print("  + Using Vertex AI.")
        if google_cloud_location:
            print(f"  - Endpoint: {google_cloud_location}-aiplatform.googleapis.com")
        else:
            print("  - GOOGLE_CLOUD_LOCATION is not set, cannot determine endpoint.")
    else:
        print("  - No API key or Vertex AI configuration found.")


def main():
    print("Gemini Doctor: Environment Diagnostics\n")
    check_node_version()
    check_cli_version()
    check_gcloud_auth()
    check_env()
    check_api_endpoint()
    print("\nDoctor checks complete.")

if __name__ == "__main__":
    main()

File: scripts/gemini-doctor.js
#!/usr/bin/env node

import fs from 'fs';
import path from 'path';
import { execSync } from 'child_process';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const GREEN = '\x1b[32m';
const RED = '\x1b[31m';
const RESET = '\x1b[0m';

function printOk(msg) {
  console.log(`${GREEN}${msg} [OK]${RESET}`);
}

function printBad(msg) {
  console.error(`${RED}${msg}${RESET}`);
}

function checkNodeVersion() {
  console.log('Checking Node.js version...');
  const requiredMajor = 20, requiredMinor = 0, requiredPatch = 0;
  const nodeVersion = process.version.replace(/^v/, '');
  const [major, minor, patch] = nodeVersion.split('.').map(Number);
  if (
    major < requiredMajor ||
    (major === requiredMajor && minor < requiredMinor) ||
    (major === requiredMajor && minor === requiredMinor && patch < requiredPatch)
  ) {
    printBad(`Node.js version ${nodeVersion} is too old. Required: >=20.0.0`);
  } else {
    printOk(`Node.js version ${nodeVersion}`);
  }
}

function checkCliVersion() {
  console.log('\nChecking Gemini CLI version...');
  try {
    const packageJsonPath = path.join(__dirname, '..', 'package.json');
    const data = JSON.parse(fs.readFileSync(packageJsonPath, 'utf8'));
    const version = data.version;
    if (version) {
      printOk(`Gemini CLI version ${version} is OK.`);
    } else {
      printBad('Gemini CLI version not found in package.json.');
    }
  } catch (err) {
    printBad('Could not read package.json: ' + err.message);
  }
}

function checkGcloudAuth() {
  console.log('\nChecking gcloud authentication...');
  const adcPath = process.env.GOOGLE_APPLICATION_CREDENTIALS;
  if (adcPath) {
    if (fs.existsSync(adcPath)) {
      printOk(`Using credentials from GOOGLE_APPLICATION_CREDENTIALS: ${adcPath}`);
      return;
    } else {
      printBad(`GOOGLE_APPLICATION_CREDENTIALS is set, but the file does not exist: ${adcPath}`);
      return;
    }
  }
  try {
    execSync('gcloud auth application-default print-access-token', { stdio: 'ignore' });
    printOk('gcloud ADC authentication is valid');
  } catch (err) {
    if (err.code === 'ENOENT') {
      printBad('gcloud command not found.');
      console.log('  - Please install the Google Cloud CLI (gcloud) and ensure it is in your PATH.');
    } else {
      printBad('gcloud ADC authentication failed.');
      console.log('  - Try running "gcloud auth application-default login" or setting GOOGLE_APPLICATION_CREDENTIALS.');
      if (err.stdout) {
        console.log('  - gcloud output:\n' + err.stdout.toString());
      }
    }
  }
}

function checkEnv() {
  console.log('\nChecking environment variables...');
  ['GOOGLE_CLOUD_PROJECT', 'GOOGLE_CLOUD_LOCATION'].forEach((varName) => {
    const value = process.env[varName];
    if (value) {
      printOk(`${varName} is set: ${value}`);
    } else {
      printBad(`${varName} is NOT set.`);
    }
  });
}

function checkApiEndpoint() {
  console.log('\nChecking API endpoint configuration...');
  const geminiApiKey = process.env.GEMINI_API_KEY;
  const googleApiKey = process.env.GOOGLE_API_KEY;
  const googleCloudProject = process.env.GOOGLE_CLOUD_PROJECT;
  const googleCloudLocation = process.env.GOOGLE_CLOUD_LOCATION;
  if (geminiApiKey) {
    console.log('  + Using Gemini API key.');
    console.log('  - Endpoint: generativelanguage.googleapis.com');
  } else if (googleApiKey || (googleCloudProject && googleCloudLocation)) {
    console.log('  + Using Vertex AI.');
    if (googleCloudLocation) {
      console.log(`  - Endpoint: ${googleCloudLocation}-aiplatform.googleapis.com`);
    } else {
      console.log('  - GOOGLE_CLOUD_LOCATION is not set, cannot determine endpoint.');
    }
  } else {
    console.log('  - No API key or Vertex AI configuration found.');
  }
}

function main() {
  console.log('Gemini Doctor: Environment Diagnostics\n');
  checkNodeVersion();
  checkCliVersion();
  checkGcloudAuth();
  checkEnv();
  checkApiEndpoint();
  console.log('\nDoctor checks complete.');
}

main();
